In [1]:
import numpy as np
from IPython.display import HTML
from bokeh.plotting import output_notebook, show
import genomes_dnj.lct_interval.series_plots as dm
import genomes_dnj.lct_interval.anal_series as an
import genomes_dnj.lct_interval.series_masks as sm
output_notebook(hide_banner=True)

<h3>117_1685 Selection Processes</h3>
<div style="width:700px">
<p>
The series 117_1685 extends across a 618,000 base region of chromosome
2 that includes the genes rab3gap1, zranb3 and half of r3hdm1.  The
435,000 base series 123_1561 extends over most of the same region.
So do the 509,000 base series 62_1265 and the 628,000 base series
193_843.  Some complex history of genetic events was required to
generate these series in the different combinations that are heavily
over expressed in 1000 genomes populations.  It is not possible to
reconstruct the details of that history.  But, some of its major features
are represented in the different associations of SNP series expressed by
1000 genomes chromosome samples.
<p>
The data presented in this notebook shows five independent hierarchies
where the series 117_1685 is over expressed.  Most chromosome samples
expressing associations of series that are part of any of those five hierarchies
express all 117 SNPs of the series 117_1685.  Three of those hierarchies selected
117_1685 along with 123_1561, and 62_1265.  Chromosome samples expressing those
hierarchies include 843 that express 193_843, 329 that express the series 67_329
and 56 that express the series 209_56.
<p>
The other two series 74_210 and 290_16 selected 117_1685 without 123_1561 or
62_1265.  The 210 chromosome samples that express the 74_210 hierarchy are
clean of both 123_1561 and 62_1265.  Those samples include 179 that express
none of the series 123_1561 SNPs and 207 that express none of the 62_1265 SNPs.
In contrast, the series 290_16 is associated with significant fragments of
both 123_1561 and 62_1265.  The chromosome samples that express 290_16 express
95 or 96 SNPs from the series 123_1561 and 40 SNPs from the series 62_1265.
<p>
A different kind of history of series SNP remodeling is illustrated by the
chromosome samples that express the series 28_434.  That history includes hierarchies
with 117_1685 on its own, 117_1685 with 123_1561, 117_1685 with 13_1696, and
123_1561 with 13_1696.  Each of the hierarchies formed with these different
associations includes fragments of series that are not fully expressed
and a variety of selector series.
<p>
The processes for generating all of these heavily over expressed complex associations
of dense SNP series remain unclear and are likely to have varied for different
hierarchies.  But, it is hard to imagine processes that could have succeeded
in all of these independent selections of 117_1685 without some functional
logic to that series of SNPs.
</div>
<h3>193_843</h3>
<div style="width:700px">
<p>
The first plot shows the most common series for chromosome samples
that express 117_1685 and 193_843.  The series 117_1685 is expressed
by 842 of the 843 chromosome samples that express 193_843.  Data on series
SNP counts shows that 702 of these chromosome samples express all 117 
series 117_1685 SNPs.  Another 103 chromosome samples express 116 of those
SNPs.
</div>

In [2]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_193_843], min_match=0.5)
plt = plt_obj.do_plot()
am_193_843 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [3]:
HTML(plt_obj.get_html())

In [4]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_193_843)
count_data

array([(109,   1), (113,   7), (114,   3), (115,  26), (116, 103),
       (117, 702)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>67_329</h3>
<div style="width:700px">
<p>
This plot shows the hierarchy selected by series 67_329, the selector expressed
by the second largest number of chromosome samples that express 117_1685.  Like 
series 193_843, 67_329 selects a combination of 117_1685, 123_1561, 
and 62_1265.  This plot's data shows that 314 of the chromosome samples that express 
67_329 also express 117_1685.  SNP count data shows that 281 of those chromosome samples 
express all 117 of the 117_1685 SNPs.  Another 24 express 116 SNPs.
</div>

In [5]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_67_329], min_match=0.5)
plt = plt_obj.do_plot()
am_67_329 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [6]:
HTML(plt_obj.get_html())

In [7]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_67_329)
count_data

array([(114,   1), (115,   8), (116,  24), (117, 281)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>74_210</h3>
<div style="width:700px">
<p>
This plot shows the hierarchy selected by the series 74_210.  All
210 of the chromosome samples that express this 74 SNP series also express
117_1685.  Neither 123_1561 nor 62_1265 are part of this hierarchy.
SNP count data shows that 109 of these chromosome samples express all 117 of
the 117_1685 SNPs.  Another 89 of the chromosome samples express 116 SNPs.
Count data for series 123_1561 SNPs shows that the chromosome samples expressing
74_210 are free of any fragments of 123_1561.  The data shows that
179 of those chromosome samples express no 123_1561 SNPs and 207 express no 62_1265 SNPs.
The chromosome samples that expess 74_210 are the largest factor in identifying 117_1685
as a unique series distinct from 123_1561 and 62_1265.
</div>

In [8]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_74_210], min_match=0.5)
plt = plt_obj.do_plot()
am_74_210 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [9]:
HTML(plt_obj.get_html())

In [10]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_74_210)
count_data

array([(112,   2), (114,   2), (115,   8), (116,  89), (117, 109)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [11]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_74_210)
count_data

array([( 0, 179), ( 1,  22), ( 2,   8), (12,   1)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [35]:
count_data = an.sa_62_1265.unique_snps_per_allele(am_74_210)
count_data

array([(0, 207), (1,   2), (8,   1)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>70_166 not 13_1696</h3>
<div style="width:700px">
<p>
The series 70_166 has selected two of the hierarchies associated with
the series 28_434.  Those two hierarchies are distinguished by the
presence or absence of the series 13_1696.  This plot shows the hierarchy
associated with 70_166 but not 13_1696.  The next plot shows the one with
13_1696.  Both plots show data for chromosome samples that express
both 70_166 and 117_1685.  The data for the first plot represents 92 chromosome
samples that express both 117_1685 and 70_166.  These chromosome samples include
62 that also express the series 49_136.  SNP count data for 117_1685 show that more
chromosome samples express 116 SNPs than express all 117.  The SNP counts for
123_1561 show that only 26 of the chromosome samples express enough of the 123
SNPs to be counted as expressing the series.  But, the rest of 92 chromosome
samples all express large fragments of 123_1561.
</div>

In [12]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_70_166], [dm.di_13_1696], min_match=0.5)
plt = plt_obj.do_plot()
am_70_166_not_13_1696 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [13]:
HTML(plt_obj.get_html())

In [14]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_70_166_not_13_1696)
count_data

array([(114,  4), (115,  8), (116, 53), (117, 27)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [15]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_70_166_not_13_1696)
count_data

array([(106,  2), (107,  4), (108,  5), (109, 38), (110, 17), (115, 24),
       (116,  2)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>70_166 and 13_1696</h3>
<div style="width:700px">
<p>
This plot shows data for 74 chromosome samples that express 70_166, 117_685, and 13_1696.
All of the 74 samples also express 28_434 and 49_136.  The 74 samples do express enough
117_1685 SNPs to be counted as expressing the series.  But, all of the samples express
numbers of 117_1685 SNPs that just meet the minimum criteria for being counted. 69
smaples express 106 of the 117_1685 SNPs.  The others express 105 or 107 of the SNPs.
The 74 chromosome samples include all 73 of the samples that express the series 22_73.
The main features of the process that lead to this hierarchy seem clear.  A recombination
event of a chromosome that expressed 70_166, 117_1685, 28_434, and 49_136 with one
that expressed 13_1696 yielded a chromosome that expressed 13_1696 and 117_1685
without the 117_1685 SNPs that overlap 13_1696.  That chromosome was selected
for overexpression by the appearance of the series 22_73.
</div>

In [16]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_70_166, dm.di_13_1696], min_match=0.5)
plt = plt_obj.do_plot()
am_70_166_and_13_1696 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [17]:
HTML(plt_obj.get_html())

In [18]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_70_166_and_13_1696)
count_data

array([(105,  3), (106, 69), (107,  2)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [39]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_70_166_and_13_1696)
count_data

array([(107,  8), (108, 50), (109, 15), (110,  1)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>209_56</h3>
<div style="width:700px">
<p>
This plot shows data for another 56 chromosome samples that express 117_1685, 123_1561,
and 62_1265.  This hierarchy has been selected by the 209 SNP series 209_56.
The strong correlations with additional series in the upper part of the region
suggests that the initial selection series included 14_48 and 80_38.  The series
26_1414 and 10_2206 are parts of the EUR tree root.  The appearance of a
different association of a number of series in a different context adds evidence
that these series have more substance than a chance association preserved
by linkage disequilibrium.
</div>

In [20]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_209_56], min_match=0.5)
plt = plt_obj.do_plot()
am_209_56 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [21]:
HTML(plt_obj.get_html())

In [22]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_209_56)
count_data

array([(116, 16), (117, 40)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [36]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_209_56)
count_data

array([(121,  7), (122, 49)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [37]:
count_data = an.sa_62_1265.unique_snps_per_allele(am_209_56)
count_data

array([(61,  2), (62, 54)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>290_16</h3>
<div style="width:700px">
<p>
This plot shows data for the 16 chromosome samples that express the 290 SNP series
290_16.  All 16 of these series express all 117 SNPs of the series 117_1685.
All of these chromosome samples also express a 95 or 96 SNP fragment of 123_1561.
</div>

In [23]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_290_16], min_match=0.5)
plt = plt_obj.do_plot()
am_290_16 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [24]:
HTML(plt_obj.get_html())

In [25]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_290_16)
count_data

array([(117, 16)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [26]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_290_16)
count_data

array([(95,  2), (96, 14)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [38]:
count_data = an.sa_62_1265.unique_snps_per_allele(am_290_16)
count_data

array([(40, 16)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>9_39</h3>
<div style="width:700px">
<p>
This plot shows data for the 39 chromosome samples that express the series
9_39.  All of these chromosome samples express an 111 or 110 SNP fragment
of 117_1685 that is large enough to be counted as an instance of that
series.  The chromosome samples also express a 95 or 96 SNP fragment of 
123_1561 that is not large enough to be counted as an instance of it.
<p>
In the case of 9_39, the history that lead to this hierarchy can
be reconstructed.  The origin of this hierarchy is in the part of
the SAS tree selected by the series 8_267.  A recombination event associated
the upper part of the the EAS tree root that included series 32_1361 and 81_857
with an instance of that hierarchy.  As a consequence of the recombination
some higher positioned SNPs of 117_1685 and 123_1561 were lost.
The appearance of the series 9_39 selected the association of series
that resulted from this recombination event and resulted in their
overexpression.
</div>

In [27]:
plt_obj = dm.superset_yes_no([dm.di_117_1685, dm.di_9_39], min_match=0.5)
plt = plt_obj.do_plot()
am_9_39 = plt_obj.plot_context.yes_allele_mask
show(plt)

In [28]:
HTML(plt_obj.get_html())

In [29]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_9_39)
count_data

array([(110,  1), (111, 38)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [30]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_290_16)
count_data

array([(95,  2), (96, 14)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

<h3>Noise</h3>
<div style="width:700px">
<p>
This plot shows data for the 42 remaining chromosomes that express 117_1685
without any association with an obvious selector series.  The 117_1685 and
123_1561 SNP counts for these chromosomes show more variation
then is generally observed when the series are associated with some
selector series.
</div>

In [31]:
plt_obj = dm.superset_yes_no([dm.di_117_1685], 
                             [dm.di_9_39, dm.di_290_16, dm.di_209_56, dm.di_70_166, dm.di_74_210, 
                              dm.di_67_329, dm.di_193_843], min_match=0.5)
plt = plt_obj.do_plot()
am_rest = plt_obj.plot_context.yes_allele_mask
show(plt)

In [32]:
HTML(plt_obj.get_html())

In [33]:
count_data = an.sa_117_1685.unique_snps_per_allele(am_rest)
count_data

array([(105, 3), (106, 1), (107, 3), (108, 6), (111, 3), (112, 1),
       (113, 2), (114, 2), (115, 5), (116, 9), (117, 7)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])

In [34]:
count_data = an.sa_123_1561.unique_snps_per_allele(am_rest)
count_data

array([(  0, 5), (  1, 1), ( 17, 6), ( 51, 1), ( 52, 2), ( 66, 1),
       ( 90, 1), ( 91, 5), (107, 3), (109, 6), (110, 1), (119, 1),
       (121, 1), (122, 4), (123, 4)], 
      dtype=[('count', '<u2'), ('snps', '<u2')])